source/extensions/http/injected_credentials/oauth2/token_provider.cc

```diff
@@ -0,0 +1,125 @@
+      oauth_scopes_(oauthScopesList(proto_config.scopes())), dispatcher_(&dispatcher),
+      stats_(generateStats(stats_prefix + "oauth2.", scope)),
+      retry_interval_(
+          proto_config.token_fetch_retry_interval().seconds() > 0
+              ? std::chrono::seconds(proto_config.token_fetch_retry_interval().seconds())
+              : std::chrono::seconds(2)) {
+  ThreadLocalOauth2ClientCredentialsTokenSharedPtr empty(
+      new ThreadLocalOauth2ClientCredentialsToken(""));
+  tls_->set(
+      [empty](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr { return empty; });
+  // initialize oauth2 client
+  oauth2_client_ = std::make_unique<OAuth2ClientImpl>(cm, proto_config.token_endpoint());
+  // set the callback for the oauth2 client
+  oauth2_client_->setCallbacks(*this);
+  asyncGetAccessToken();
+}
+
+// TokenProvider asyncGetAccessToken
+void TokenProvider::asyncGetAccessToken() {
+  // get the access token from the oauth2 client
+  if (timer_) {
+    timer_->disableTimer();
+    timer_.reset();
+  }
+  if (secret_reader_->credential().empty()) {
+    ENVOY_LOG(error, "asyncGetAccessToken: client secret is empty, retrying in {} seconds.",
+              retry_interval_.count());
+    timer_ = dispatcher_->createTimer([this]() -> void { asyncGetAccessToken(); });
+    timer_->enableTimer(std::chrono::seconds(retry_interval_));
+    stats_.token_fetch_failed_on_client_secret_.inc();
+    return;
+  }
+  auto result =
+      oauth2_client_->asyncGetAccessToken(client_id_, secret_reader_->credential(), oauth_scopes_);
+  if (result == OAuth2Client::GetTokenResult::NotDispatchedClusterNotFound) {
+    ENVOY_LOG(error, "asyncGetAccessToken: OAuth cluster not found. Retrying in {} seconds.",
+              retry_interval_.count());
+    timer_ = dispatcher_->createTimer([this]() -> void { asyncGetAccessToken(); });
+    timer_->enableTimer(std::chrono::seconds(retry_interval_));
+    stats_.token_fetch_failed_on_cluster_not_found_.inc();
+    return;
+  }
+  stats_.token_requested_.inc();
+  ENVOY_LOG(debug, "asyncGetAccessToken: Dispatched OAuth request for access token.");
+}
+
+// FilterCallbacks
+void TokenProvider::onGetAccessTokenSuccess(const std::string& access_token,
+                                            std::chrono::seconds expires_in) {
+  // set the token
+  auto token = absl::StrCat("Bearer ", access_token);
+  ThreadLocalOauth2ClientCredentialsTokenSharedPtr value(
+      new ThreadLocalOauth2ClientCredentialsToken(token));
+  stats_.token_fetched_.inc();
+
+  tls_->set(
+      [value](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr { return value; });
+  if (timer_) {
+    return;
+  }
+  timer_ = dispatcher_->createTimer([this]() -> void { asyncGetAccessToken(); });
+  ENVOY_LOG(debug, "onGetAccessTokenSuccess: Token fetched successfully, expires in {} seconds.",
+            expires_in.count());
+  timer_->enableTimer(expires_in / 2);
```

Comment:
I think the things about timer could be simper. You completely needn't create the timer again and again. You can only create one and enable/disable it according your requirement here.